parser-typechecker/src/U/Codebase/Branch/Diff.hs

```diff
@@ -0,0 +1,169 @@
+import qualified Data.Set as Set
+import Data.These
+import U.Codebase.Branch
+import qualified U.Codebase.Branch.Type as Branch
+import qualified U.Codebase.Causal as Causal
+import U.Codebase.Reference (Reference)
+import U.Codebase.Referent (Referent)
+import Unison.Name (Name)
+import qualified Unison.Name as Name
+import qualified Unison.NameSegment as NameSegment
+import Unison.Prelude
+
+data Diff a = Diff
+  { adds :: Set a,
+    removals :: Set a
+  }
+  deriving (Show, Eq, Ord)
+
+-- | Represents the changes to definitions at a given path, not including child paths.
+--
+-- Note: doesn't yet include any info on metadata or patch diffs. Feel free to add it.
+data DefinitionDiffs = DefinitionDiffs
+  { termDiffs :: Map NameSegment (Diff Referent),
+    typeDiffs :: Map NameSegment (Diff Reference)
+    -- termMetadataDiffs :: Map (NameSegment, Referent) (Diff Reference),
+    -- typeMetadataDiffs :: Map (NameSegment, Reference) (Diff Reference)
+    -- patchDiffs :: Map NameSegment (Diff ())
+  }
+  deriving stock (Show, Eq, Ord)
+
+instance Semigroup DefinitionDiffs where
+  a <> b =
+    DefinitionDiffs
+      { termDiffs = termDiffs a <> termDiffs b,
+        typeDiffs = typeDiffs a <> typeDiffs b
+      }
+
+instance Monoid DefinitionDiffs where
+  mempty = DefinitionDiffs mempty mempty
+
+-- | A tree of local diffs. Each node of the tree contains the definition diffs at that path.
+newtype TreeDiff = TreeDiff
+  { unTreeDiff :: Cofree (Map NameSegment) DefinitionDiffs
+  }
+  deriving stock (Show, Eq, Ord)
+
+instance Semigroup TreeDiff where
+  TreeDiff (a :< as) <> TreeDiff (b :< bs) =
+    TreeDiff $ (a <> b) :< (Map.unionWith mergeCofrees as bs)
+    where
+      mergeCofrees x y = unTreeDiff (TreeDiff x <> TreeDiff y)
+
+instance Monoid TreeDiff where
+  mempty = TreeDiff (mempty :< mempty)
+
+instance Lens.AsEmpty TreeDiff where
+  _Empty = Lens.only mempty
+
+-- | A summary of a 'TreeDiff', containing all names added and removed.
+-- Note that there isn't a clear notion of a name "changing" since conflicts might muddy the notion
+-- by having multiple copies of both the from and to names, so we just talk about adds and
+-- removals instead.
+data NameChanges = NameChanges
+  { termNameAdds :: [(Name, Referent)],
+    termNameRemovals :: [(Name, Referent)],
+    typeNameAdds :: [(Name, Reference)],
+    typeNameRemovals :: [(Name, Reference)]
+  }
+
+instance Semigroup NameChanges where
+  (NameChanges a b c d) <> (NameChanges a2 b2 c2 d2) =
+    NameChanges (a <> a2) (b <> b2) (c <> c2) (d <> d2)
+
+instance Monoid NameChanges where
+  mempty = NameChanges mempty mempty mempty mempty
+
+-- | Diff two Branches, returning a tree containing all of the changes
+diffBranches :: forall m. Monad m => Branch m -> Branch m -> m TreeDiff
+diffBranches from to = do
+  let termDiffs = diffMap (terms from) (terms to)
+  let typeDiffs = diffMap (types from) (types to)
+  let defDiff = DefinitionDiffs {termDiffs, typeDiffs}
+  rec <- do
```

Comment:
Lame suggestion: maybe rename this variable? `rec` is a keyword when `RecursiveDo` is enabled, so it gets special treatment by regexy syntax highlighters.